README.md
# Challenge Técnico Telecom - Eduardo Sanchez Lopez

## Instrucciones para levantar el servicio
 - copiar [este](https://drive.google.com/file/d/1PqRGceIOO4Akg3erhKH0RcyAvgHIsQci/view?usp=sharing) archivo .env en la ruta raiz del proyecto
 - ```yarn install``` o ```npm install``` para instalar las dependecias del proyecto
 - ```yarn install``` o ```npm run start``` para iniciar el proyecto

## Instrucciones para levantar el servicio usando docker-compose
- copiar [este](https://drive.google.com/file/d/1PqRGceIOO4Akg3erhKH0RcyAvgHIsQci/view?usp=sharing) archivo .env en la ruta raiz del proyecto
 - ```docker-compose up --build -d``` para levantar los servicios definidos en **docker-compose.yml**
 - ```docker-compose down``` para detener todos los servicios relacionados al proyecto
> Nota: si se utiliza `docker` la ubicacion por ip del request no puede ser obtenida en el entorno local, tiene un valor mock.

### Otros comandos utiles
- ```yarn lint``` o ```npm run lint``` para correr el analsis estatico del codigo
- ```yarn test``` o ```npm run test``` para correr los test que se encuentran en **./test**
- ```yarn release``` o ```npm run release``` para lanzar una nueva version en el changelog
- ```yarn dev``` o ```npm run dev``` para iniciar el servicio en modo desarrollo

## Documentacion de los endpoints:
 - [Postman](https://documenter.getpostman.com/view/17908890/2s83f8isGn)


CHANGELOG.md
# Changelog

All notable changes to this project will be documented in this file. See [standard-version](https://github.com/conventional-changelog/standard-version) for commit guidelines.

## [1.5.0](https://github.com/WASD6570/teco/compare/v1.4.1...v1.5.0) (2022-09-28)


### Features

* prodscript ([16c7984](https://github.com/WASD6570/teco/commit/16c7984f029a4d652212b0a4ccf5e9ae063fd5af))

## [1.4.0](https://github.com/WASD6570/teco/compare/v1.3.0...v1.4.0) (2022-09-25)


### Features

* add docker ([3e8a3a1](https://github.com/WASD6570/teco/commit/3e8a3a1084748fa61a7df413e7ae017f7d6bf328))


### Bug Fixes

* check private ip ranges in get_ip.js ([1a3eb71](https://github.com/WASD6570/teco/commit/1a3eb71e89f17a8cbcc0ca0f1d36e46d41ebab0d))

## [1.3.0](https://github.com/WASD6570/teco/compare/v1.2.0...v1.3.0) (2022-09-25)


### Features

* add husky ([dcb7b67](https://github.com/WASD6570/teco/commit/dcb7b676f5ff59799d8ebfb2cea992ec553618f5))
* add test and coverage ([0a0b511](https://github.com/WASD6570/teco/commit/0a0b511be15f3810b1467e598cd7776fac2aa654))

## [1.2.0](https://github.com/WASD6570/teco/compare/v1.1.0...v1.2.0) (2022-09-25)


### Features

* add eslint ([e67d321](https://github.com/WASD6570/teco/commit/e67d321b5a83bb348b1acce192032c67996dcca8))

## 1.1.0 (2022-09-25)


### Features

* initial commit ([b520b4c](https://github.com/WASD6570/teco/commit/b520b4ce15af855ebf56c213e5cb59d34b0f63b8))


tests/get_ip.test.js
const { getIp } = require("../src/services/get_ip.js");

test("It should return the server IPv4", async () => {
  const ip = await getIp({ ip: "::1" });
  expect(ip).toEqual(
    expect.stringMatching(
      /^((25[0-5]|(2[0-4]|1[0-9]|[1-9]|)[0-9])(\.(?!$)|$)){4}$/
    )
  );
});

test("It should return the mock IPv4", async () => {
  const ip = await getIp({ ip: "200.11.22.333" });
  expect(ip).toEqual("200.11.22.333");
});


tests/weather.test.js
const { getWeather } = require("../src/services/weather");

test("It should return the current weather of the city", async () => {
  const weather = await getWeather({ city: "London" });
  expect(weather).toHaveProperty("current");
  expect(weather).toHaveProperty("location");
  expect(weather.location).toHaveProperty("region");
  expect(weather.location.region).toBe("City of London, Greater London");
});

test("It should return teh current weather of the mock lat-lon pair", async () => {
  const weather = await getWeather({ lat: 51.5074, lon: 0.1278 });
  expect(weather).toHaveProperty("current");
  expect(weather).toHaveProperty("location");
  expect(weather.location).toHaveProperty("region");
  expect(weather.location.region).toBe("Bexley, Greater London");
});

test("It should return the forecast weather of the city", async () => {
  const weather = await getWeather({
    city: "London",
    method: "forecast",
  });
  expect(weather).toHaveProperty("forecast");
  expect(weather).toHaveProperty("location");
  expect(weather.location).toHaveProperty("region");
  expect(weather.location.region).toBe("City of London, Greater London");
  expect(weather.forecast).toHaveProperty("forecastday");
});

test("It should return the forecast weather of the city", async () => {
  const weather = await getWeather({
    lat: 51.5074,
    lon: 0.1278,
    method: "forecast",
  });
  expect(weather).toHaveProperty("forecast");
  expect(weather).toHaveProperty("location");
  expect(weather.location).toHaveProperty("region");
  expect(weather.location.region).toBe("Bexley, Greater London");
  expect(weather.forecast).toHaveProperty("forecastday");
});


tests/location.test.js
const { getLocation } = require("../src/services/location");

test("It should return the location of the ip address", async () => {
  const location = await getLocation({ ip: "200.59.81.129" });
  expect(location).toHaveProperty("status");
  expect(location.status).toBe("success");
  expect(location.regionName).toBe("San Juan");
});


src/routes/forecast.js
const { getLocation } = require("../services/location");
const { weather_api_key, weather_api_url } = require("../config/default.js");
const fetch = require("../services/fetch.js");

/**
 * @swagger
 * tags:
 *  name: forecast
 *  description: forecast endpoints
 */

/**
 * @swagger
 * /forecast/{city}:
 *  get:
 *   summary: Get forecast weather by city name
 *  tags:
 *   - Forecast
 *  externalDocs:
 *    description: Find out more about this endpoints
 *    url: https://documenter.getpostman.com/view/17908890/2s83f8isGn
 *  parameters:
 *    - in: path
 *    name: city
 *    schema:
 *      type: string
 *      required: true
 *      description: City name
 *    - in: query
 *    name: days
 *    schema:
 *      type: integer
 *      required: false
 *      description: Number of days requested, default 5
 *  responses:
 *    200:
 *     description: Forecast weather
 *     content:
 *      application/json:
 *    404:
 *      description: Error getting forecast weather
 *      content:
 *        application/json:
 *        schema:
 *          type: object
 *          properties:
 *            error:
 *            type: string
 *    500:
 *      description: Error getting forecast weather
 *      content:
 *        application/json:
 *        schema:
 *          type: object
 *          properties:
 *            error:
 *            type: string
 */

async function forecastRoutes(fastify) {
  fastify.get("/forecast/:city", async (request, reply) => {
    try {
      const city = request.params.city;
      const days = request.query.days || 5;
      const data = await fetch(
        `${weather_api_url}/forecast.json?key=${weather_api_key}&q=${city}&days=${days}`
      );
      const json = await data.json();
      if (!json.error) {
        return reply.send({ location: json.location, forecast: json.forecast });
      }
      reply.code(404).send({ error: json.error.message });
    } catch (error) {
      reply.code(500).send({ error: "Error getting forecast wheather" });
    }
  });
  /**
   * @swagger
   * /forecast
   *  get:
   *   summary: Get forecast weather of the client location
   *  tags:
   *   - Forecast
   *  externalDocs:
   *    description: Find out more about this endpoints
   *    url: https://documenter.getpostman.com/view/17908890/2s83f8isGn
   *  parameters:
   *    - in: query
   *    name: days
   *    schema:
   *      type: integer
   *      required: false
   *      description: Number of days requested, default 5
   *  responses:
   *    200:
   *     description: Forecast weather
   *     content:
   *      application/json:
   *    404:
   *      description: Error getting forecast weather
   *      content:
   *        application/json:
   *        schema:
   *          type: object
   *          properties:
   *            error:
   *            type: string
   *    500:
   *      description: Error getting forecast weather
   *      content:
   *        application/json:
   *        schema:
   *          type: object
   *          properties:
   *            error:
   *            type: string
   */

  fastify.get("/forecast", async (request, reply) => {
    try {
      const days = request.query.days || 5;
      const location = await getLocation(request);
      const data = await fetch(
        `${weather_api_url}/forecast.json?key=${weather_api_key}&q=${location.lat},${location.lon}&days=${days}`
      );
      const json = await data.json();
      if (!json.error) {
        return reply.send({ location: json.location, forecast: json.forecast });
      }
      reply.code(404).send({ error: json.error.message });
    } catch (error) {
      reply.code(500).send({ error: "Error getting forecast wheather" });
    }
  });
}

module.exports = forecastRoutes;


src/routes/location.js
const { getLocation } = require("../services/location");

/**
 * @swagger
 * tags:
 *  name: location
 *  description: location endpoint
 */

/**
 * @swagger
 * /location
 *   get:
 *     summary: Get current location by ip
 *     tags:
 *       - Location
 *     externalDocs:
 *       description: Find out more about this endpoints
 *       url: https://documenter.getpostman.com/view/17908890/2s83f8isGn
 *     responses:
 *       200:
 *         description: Current location
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                  status: string
 *                  country: string
 *                  countryCode: string
 *                  region: string
 *                  regionName: string
 *                  city: string
 *                  zip: string
 *                  lat: number
 *                  lon: number
 *                  timezone: string
 *                  isp: string
 *                  org: string
 *                  as: string
 *                  query: string
 *       500:
 *         description: Error getting location
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 error:
 *                   type: string
 */

async function locationRoutes(fastify) {
  fastify.get("/location", async (request, reply) => {
    try {
      const location = await getLocation(request);
      reply.send(location);
    } catch (error) {
      reply.code(500).send({ error: "Error getting location" });
    }
  });
}

module.exports = locationRoutes;


src/services/weather.js
const fetch = require("./fetch");
const { weather_api_key, weather_api_url } = require("../config/default.js");

const getWeather = async ({ method = "current", city, days, lat, lon }) => {
  let data;
  if (city !== undefined && days !== undefined && method === "forecast") {
    data = await fetch(
      `${weather_api_url}/${method}.json?key=${weather_api_key}&q=${city}&days=${days}`
    );
  }
  if (city !== undefined) {
    data = await fetch(
      `${weather_api_url}/${method}.json?key=${weather_api_key}&q=${city}`
    );
  }
  if (lat !== undefined && lon !== undefined) {
    data = await fetch(
      `${weather_api_url}/${method}.json?key=${weather_api_key}&q=${lat},${lon}`
    );
  }
  if (
    lat !== undefined &&
    lon !== undefined &&
    days !== undefined &&
    method === "forecast"
  ) {
    data = await fetch(
      `${weather_api_url}/${method}.json?key=${weather_api_key}&q=${lat},${lon}&days=${days}`
    );
  }
  return await data?.json();
};

module.exports = { getWeather };


src/routes/current.js
const { getLocation } = require("../services/location");
const { getWeather } = require("../services/weather.js");

/**
 * @swagger
 * tags:
 *  name: current
 *  description: current endpoints
 */

/**
 * @swagger
 * /current/{city}:
 *   get:
 *     summary: Get current weather by city name
 *     tags:
 *       - Current
 *     externalDocs:
 *       description: Find out more about this endpoints
 *       url: https://documenter.getpostman.com/view/17908890/2s83f8isGn
 *     parameters:
 *       - in: path
 *         name: city
 *         schema:
 *           type: string
 *         required: true
 *         description: City name
 *     responses:
 *       200:
 *         description: Current weather
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                  status: string
 *                  country: string
 *                  countryCode: string
 *                  region: string
 *                  regionName: string
 *                  city: string
 *                  zip: string
 *                  lat: number
 *                  lon: number
 *                  timezone: string
 *                  isp: string
 *                  org: string
 *                  as: string
 *                  query: string
 *       404:
 *          description: Error getting current weather
 *             content:
 *                application/json:
 *                  schema:
 *                    type: object
 *                    properties:
 *                      error:
 *                        type: string
 *       500:
 *         description: Error getting current weather
 *         content:
 *           application/json:
 *             schema:
 *               type: object
 *               properties:
 *                 error:
 *                   type: string
 */

async function currentRoutes(fastify) {
  fastify.get("/current/:city", async (request, reply) => {
    try {
      const city = request.params.city;
      const json = await getWeather({ method: "current", city });
      if (!json.error) {
        return reply.send({ location: json.location, current: json.current });
      }
      return reply.status(404).send({ error: json.error.message });
    } catch (error) {
      reply.code(500).send({ error: "Error getting current wheather" });
    }
  });

  /**
   * @swagger
   * /current
   *    get:
   *      summary: Get current weather by location of the client
   *    tags:
   *      - Current
   *    externalDocs:
   *      description: Find out more about this endpoints
   *      url: https://documenter.getpostman.com/view/17908890/2s83f8isGn
   *    responses:
   *      200:
   *        description: Current weather
   *        content:
   *           application/json:
   *             schema:
   *               type: object
   *               properties:
   *                  status: string
   *                  country: string
   *                  countryCode: string
   *                  region: string
   *                  regionName: string
   *                  city: string
   *                  zip: string
   *                  lat: number
   *                  lon: number
   *                  timezone: string
   *                  isp: string
   *                  org: string
   *                  as: string
   *                  query: string
   *      404:
   *        description: Error getting current weather
   *         content:
   *                application/json:
   *                  schema:
   *                    type: object
   *                    properties:
   *                      error:
   *                        type: string
   *      500:
   *        description: Error getting current weather
   *        content:
   *            application/json:
   *              schema:
   *                type: object
   *                properties:
   *                  error:
   *                    type: string
   */

  fastify.get("/current", async (request, reply) => {
    try {
      const location = await getLocation(request);
      const json = await getWeather({
        method: "current",
        lat: location.lat,
        lon: location.lon,
      });
      if (!json.error) {
        return reply.send({ location: json.location, current: json.current });
      }
      reply.status(404).send({ error: json.error.message });
    } catch (error) {
      reply.code(500).send({ error: "Error getting current wheather" });
    }
  });
}

module.exports = currentRoutes;
